Feature: Added 'compress' attribute option to the properties window

d2dyno1 opened the pull request:
**Summary**
This PR continues on the work made in the [ya/CompressedAttribute](https://github.com/files-community/Files/tree/ya/CompressedAttribute) branch and fixes the broken logic to actually enable/disable compression for items

FYI @yaira2 

**Resolved / Related Issues**
- Closes [#15976](https://github.com/files-community/Files/issues/15976)

**Screenshots**
![image](https://github.com/user-attachments/assets/991f3897-2f39-40b8-8753-4fe47c4eca43)

**Todo**

- [x] Tested setting compression for files
- [x] Tested setting compression for folders
- [x] Add a GetVolumeInformation check to see if volume supports compression through `FILE_FILE_COMPRESSION` attribute

0x5bfa commented on src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:
```diff
@@ -935,6 +937,40 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
+		{
+			using var hFile = PInvoke.CreateFile(
+				lpFileName,
+				Win32PInvoke.GENERIC_READ | Win32PInvoke.GENERIC_WRITE,
```
```suggestion
				(uint)FILE_ACCESS_RIGHTS.FILE_WRITE_ATTRIBUES,
```

0x5bfa replied:
If it doesn't work well, let me know. I think standard write access works as well.

0x5bfa commented on src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:
```diff
@@ -935,6 +937,40 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
+			if (hFile.IsInvalid)
+				return false;
+
+			var bytesReturned = 0u;
```
```suggestion
			uint bytesReturned = 0u;
```

0x5bfa commented on src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:
```diff
@@ -935,6 +937,40 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
+				return false;
+
+			var bytesReturned = 0u;
+			var compressionFormat = isCompressed
+				? Win32PInvoke.COMPRESSION_FORMAT_DEFAULT
+				: Win32PInvoke.COMPRESSION_FORMAT_NONE;
```
```suggestion
			COMPRESSION_FORMAT compressionFormat = isCompressed
				? COMPRESSION_FORMAT.COMPRESSION_FORMAT_DEFAULT
				: COMPRESSION_FORMAT.COMPRESSION_FORMAT_NONE;
```

0x5bfa commented on src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:
```diff
@@ -935,6 +937,40 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
+				0u,
+				&bytesReturned);
+
+			if (result.Value == 0)
```
BOOL is a struct and has implicit operator overloads for conversion between BOOL and bool.
```suggestion
			if (!result)
```

0x5bfa commented on src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:
```diff
@@ -935,6 +937,40 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
+
+			var result = PInvoke.DeviceIoControl(
+				new(hFile.DangerousGetHandle()),
+				Win32PInvoke.FSCTL_SET_COMPRESSION,
```
```suggestion
				(uint)FSCTL_SET_COMPRESSION,
```

0x5bfa commented on src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:
```diff
@@ -13,6 +13,8 @@
 using Vanara.PInvoke;
 using Windows.System;
+using Windows.Win32;
+using Windows.Win32.Storage.FileSystem;
```
```suggestion
using Windows.Win32.Storage.FileSystem;
using Windows.Win32.System.Ioctl;
```

0x5bfa commented on src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:
```diff
@@ -935,6 +937,40 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
+				Win32PInvoke.FSCTL_SET_COMPRESSION,
+				&compressionFormat,
+				sizeof(ushort),
+				(void*)IntPtr.Zero,
```
I believe "null" works here.
```suggestion
				null,
```

0x5bfa commented on src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs:
Please revert changes in this file.

d2dyno1 replied:
No (see comment)

0x5bfa commented on src/Files.App/ViewModels/Properties/Items/CombinedProperties.cs:
```diff
@@ -170,6 +170,28 @@ private void ViewModel_PropertyChanged(object sender, System.ComponentModel.Prop
+
+				case "IsContentCompressed":
+					{
+						if (ViewModel.IsContentCompressed is not null)
+						{
+							if (ViewModel.IsContentCompressed ?? false)
+							{
+								List.ForEach(x =>
+								{
+									Win32Helper.SetCompressionAttributeIoctl(x.ItemPath, true);
+								});
+							}
+							else
+							{
+								List.ForEach(x =>
+								{
+									Win32Helper.SetCompressionAttributeIoctl(x.ItemPath, false);
+								});
+							}
+						}
```
```suggestion
						if (ViewModel.IsContentCompressed ?? false)
						{
							List.ForEach(x =>
							{
								Win32Helper.SetCompressionAttributeIoctl(x.ItemPath, true);
							});
						}
						else
						{
							List.ForEach(x =>
							{
								Win32Helper.SetCompressionAttributeIoctl(x.ItemPath, false);
							});
						}
```

0x5bfa commented on src/Files.App/ViewModels/Properties/Items/FileProperties.cs:
```diff
@@ -279,13 +281,23 @@ private async void ViewModel_PropertyChanged(object sender, System.ComponentMode
 					break;
 
+				case nameof(ViewModel.IsContentCompressed):
+					if (ViewModel.IsContentCompressed is not null)
+					{
+						if ((bool)ViewModel.IsContentCompressed)
+							Win32Helper.SetCompressionAttributeIoctl(Item.ItemPath, true);
+						else
+							Win32Helper.SetCompressionAttributeIoctl(Item.ItemPath, false);
+					}
```
```suggestion
					Win32Helper.SetCompressionAttributeIoctl(Item.ItemPath, ViewModel.IsContentCompressed ?? false);
```

0x5bfa commented on src/Files.App/ViewModels/Properties/Items/FolderProperties.cs:
```diff
@@ -211,6 +214,16 @@ private async void ViewModel_PropertyChanged(object sender, System.ComponentMode
 					break;
 
+				case "IsContentCompressed":
+					if (ViewModel.IsContentCompressed is not null)
+					{
+						if ((bool)ViewModel.IsContentCompressed)
+							Win32Helper.SetCompressionAttributeIoctl(Item.ItemPath, true);
+						else
+							Win32Helper.SetCompressionAttributeIoctl(Item.ItemPath, false);
+					}
```
```suggestion
					Win32Helper.SetCompressionAttributeIoctl(Item.ItemPath, ViewModel.IsContentCompressed ?? false);
```

0x5bfa commented on src/Files.App/ViewModels/Properties/Items/LibraryProperties.cs:
```diff
@@ -161,6 +162,16 @@ private void ViewModel_PropertyChanged(object sender, System.ComponentModel.Prop
 					break;
+
+				case "IsContentCompressed":
+					if (ViewModel.IsContentCompressed is not null)
+					{
+						if ((bool)ViewModel.IsContentCompressed)
+							Win32Helper.SetCompressionAttributeIoctl(Library.ItemPath, true);
+						else
+							Win32Helper.SetCompressionAttributeIoctl(Library.ItemPath, false);
+					}
```
```suggestion
					Win32Helper.SetCompressionAttributeIoctl(Item.ItemPath, ViewModel.IsContentCompressed ?? false);
```

0x5bfa reviewed (commented):
**Summary**
- Please use the constants win32metadata defines.
- Simplified some if-else's.

d2dyno1 commented:
@0x5bfa I implemented the changes you requested. Unfortunately, some of them didn't compile and CsWin32 didn't feel like SG-ing enums for me so I reverted to the old values.

0x5bfa commented:
Can you try to add "COMPRESSION_FORMAT" and "FILE_ACCESS_RIGHTS" in NativeMethods.txt beforehand and use these enums?

yaira2 commented:
@d2dyno1 fyi

0x5bfa commented on src/Files.App/Data/Models/SelectedItemsPropertiesViewModel.cs:
```diff
@@ -718,6 +718,31 @@ public bool? IsHiddenEditedValue
 		}
 
+		private bool? isContentCompressed;
+		public bool? IsContentCompressed
```
Could you add a comment like "This is only for NTFS" so others wouldn't be confused it with archive files.

0x5bfa commented on src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs:
```diff
@@ -45,8 +40,18 @@ public static partial class Win32PInvoke
+
+		// Compression
+		public const ushort COMPRESSION_FORMAT_NONE = 0x0000;
+		public const ushort COMPRESSION_FORMAT_DEFAULT = 0x0001;
```
You can use COMPRESSION_FORMAT enum.

yaira2 replied:
This should be resolved with 31289160e0e62cb693cbe5d85fc6026355b58fc7

0x5bfa commented on src/Files.App/Helpers/Win32/Win32PInvoke.Consts.cs:
```diff
@@ -45,8 +40,18 @@ public static partial class Win32PInvoke
 		public const int FSCTL_GET_REPARSE_POINT = 0x000900A8;
+		public const int FSCTL_LOCK_VOLUME = 0x00090018;
+		public const int FSCTL_DISMOUNT_VOLUME = 0x00090020;
+		public const uint FSCTL_SET_COMPRESSION = 0x9C040u;
```
I think you can add these const names to CsWin32 and use them from the `Windows.Win32.PInvoke` class like `PInvoke.FSCTL_SET_COMPRESSION`.

d2dyno1 replied:
What's the enum name?

0x5bfa replied:
As above.

yaira2 replied:
This should be resolved with 31289160e0e62cb693cbe5d85fc6026355b58fc7

0x5bfa commented on src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:
```diff
@@ -11,8 +11,16 @@
 using Windows.System;
+using Windows.Win32;
+using Windows.Win32.Foundation;
+using Windows.Win32.Storage.FileSystem;
```
If you struggled with co-existence of CsWin32 and Vanara, I don't mind to create a new file `Win32Helper.CsWin32.cs` that only contains CsWin32 references so we can do replacement of Vanara easily. However, as it depends upon you, I'll leave it up to you.

0x5bfa commented on src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:
```diff
@@ -935,10 +943,72 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
+		{
+			path = Path.GetPathRoot(path) ?? string.Empty;
+
+			// MAX_PATH = 260 + 1 (null termination)
```
`MAX_PATH` is 260 including null termination char (`\0`), and 3 (`C:\`) + 256 (= 259) can be used as path chars.

```c
// Definition
#define MAX_PATH 260

// Usage ("" of C/C++/C# appends \0 automatically and if you use all 260, this'd stack-overflow)
char szPath[MAX_PATH] = "C:\\Windows";
```

```suggestion
			// MAX_PATH = 260 (includes null termination char)
```

d2dyno1 replied:
Good to know, thanks!

0x5bfa commented on src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:
```diff
@@ -935,10 +943,72 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
+				return false;
+
+			// 0x00000010 FILE_FILE_COMPRESSION
+			return (fileSystemFlags & 0x00000010) != 0;
```
You can add FILE_FILE_COMPRESSION to CsWin32 and use from `PInvoke.FILE_FILE_COMPRESSION`

d2dyno1 replied:
What's the enum name?

0x5bfa replied:
This is not a member of an enum but a **constant** value in C.
You can add FILE_FILE_COMPRESSION and use it like PInvoke.FILE_FILE_COMPRESSION

yaira2 replied:
This should be resolved with 31289160e0e62cb693cbe5d85fc6026355b58fc7

0x5bfa commented on src/Files.App/Helpers/Win32/Win32Helper.Storage.cs:
```diff
@@ -935,10 +943,72 @@ public static bool UnsetFileAttribute(string lpFileName, FileAttributes dwAttrs)
+
+			// MAX_PATH = 260 + 1 (null termination)
+			var volumeName = new PWSTR();
+			var fileSystemName = new PWSTR();
```
Don't you have to allocate? If so, you can do as follows:

```c#
const int MAX_PATH = 260;
PWTSR pVolumeName = stackalloc char[MAX_PATH]; // you may need casting maybe?
PWSTR pFileSystemName = stackalloc char[MAX_PATH];
```

0x5bfa replied:
In that cause, I think you can specify null to the dedicated parameter.

```c#
PInvoke.GetVolumeInformation(
    path,
    null,
    0u,
    null,
    null,
    &dwFileSystemFlags,
    null,
    0u);
```

d2dyno1 replied:
Addressed in [`c50e313` (#16319)](https://github.com/files-community/Files/pull/16319/commits/c50e31335f13669d3619e4de79a20032a537160b)

d2dyno1 commented:
I think we can move on with it @0x5bfa

0x5bfa commented:
Looks great.
